File: golang/3206_number_of_alternating_groups/number_of_alternating_groups.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func numberOfAlternatingGroups(colors []int) int {
	n := len(colors)
	res := 0
	for i := 0; i < n; i++ {
		if colors[i] != colors[(i-1+n)%n] && colors[i] != colors[(i+1)%n] {
			res++
		}
	}
	return res
}

func main() {
	tests := []struct {
		colors []int
		ans    int
	}{
		{[]int{1, 1, 1}, 0},
		{[]int{0, 1, 0, 0, 1}, 3},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, numberOfAlternatingGroups(test.colors), index)
	}
}

File: cpp/3206_number_of_alternating_groups/number_of_alternating_groups.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int numberOfAlternatingGroups(vector<int>& colors) {
    int n = colors.size();
    int res = 0;
    for (int i = 0; i < n; i++) {
      if (colors[i] != colors[(i - 1 + n) % n] && colors[i] != colors[(i + 1) % n]) {
        res += 1;
      }
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<int>, int>> tests{
      {{1, 1, 1}, 0},
      {{0, 1, 0, 0, 1}, 3},
  };

  for (auto &[colors, ans] : tests) {
    assert(Solution().numberOfAlternatingGroups(colors) == ans);
  }
}